rclpy/rclpy/node.py

```diff
@@ -493,7 +489,9 @@ def declare_parameters(
             raise_on_failure=True,
             allow_undeclared_parameters=True
         )
-        return self.get_parameters([parameter.name for parameter in parameter_list])
+        # Don't call get_parameters() to bypass check for NOT_SET parameters
+        parameter_names = [parameter.name for parameter in parameter_list]
+        return [self._parameters[name] for name in parameter_names]
```

Comment:
I'm 95% sure this is okay (ie. we aren't losing anything from not calling `get_parameters`), but it deserves extra attention :sweat_smile:

Reply:
nit:

```suggestion
        return [self._parameters[parameter.name] for parameter in parameter_list]
```

Reply:
This looks ok to me, but for sure others also look at it.

I think the apprehension comes from duplicating the logic slightly. Maybe a different way to do it would be to have a function that is called here for each parameter, which is also called from `self.get_parameter()`, just so they go through the same path.

My nascent concern is that we change `get_parameter` in the future, but not this code in some breaking way, because the future author didn't realize it could be accessed from more than one place.

Reply:
I've applied Ivan's suggestion (5aeceba).

I'm not sure we can reasonably add a common function, since `get_parameters` calls `get_parameter` for each name and does some validation (checks if declared and initialized). I think trying use a common function would make the implementation too convoluted .

Reply:
> I'm not sure we can reasonably add a common function, since get_parameters calls get_parameter for each name and does some validation (checks if declared and initialized). I think trying use a common function would make the implementation too convoluted .

Ok, that's fine with me. To be clear, I was thinking a common function between `get_parameter` and this method, I was thinking `get_parameters` would still call `get_parameter`.

As I said, it was just an idea to address what I thought you were concerned about as well (we're essentially duplicating the accessor code without the validations). But I don't think it's that big of a deal to leave it as-is.